engine/consensus/approvals/assignment_collector.go

```diff
@@ -0,0 +1,346 @@
+// hangs far enough behind finalization (measured in finalized but unsealed blocks), emergency
+// sealing kicks in. This will be removed when implementation of seal & verification is finished.
+func (rsr *AssignmentCollector) emergencySealable(collector *ApprovalCollector, finalizedBlockHeight uint64) (bool, error) {
+	// Criterion for emergency sealing:
+	// there must be at least DefaultEmergencySealingThreshold number of blocks between
+	// the block that _incorporates_ result and the latest finalized block
+	return collector.IncorporatedBlock().Height+sealing.DefaultEmergencySealingThreshold <= finalizedBlockHeight, nil
+}
+
+func (rsr *AssignmentCollector) CheckEmergencySealing(finalizedBlockHeight uint64) error {
+	for _, collector := range rsr.allCollectors() {
+		sealable, err := rsr.emergencySealable(collector, finalizedBlockHeight)
+		if err != nil {
+			return fmt.Errorf("could not determnine if result %x incorporated at %x is emergency sealable: %w",
+				rsr.ResultID, collector.IncorporatedBlockID(), err)
+		}
+		if sealable {
+			err = collector.SealResult()
+			if err != nil {
+				return fmt.Errorf("could not create emergency seal for result %x incorporated at %x: %w",
+					rsr.ResultID, collector.IncorporatedBlockID(), err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (rsr *AssignmentCollector) ProcessIncorporatedResult(incorporatedResult *flow.IncorporatedResult) error {
+	// check that result is the one that this AssignmentCollector manages
+	if irID := incorporatedResult.Result.ID(); irID != rsr.ResultID {
+		return fmt.Errorf("this AssignmentCollector manages result %x but got %x", rsr.ResultID, irID)
+	}
+
+	incorporatedBlockID := incorporatedResult.IncorporatedBlockID
+	if collector := rsr.collectorByBlockID(incorporatedBlockID); collector != nil {
+		return nil
+	}
+
+	// This function is not exactly thread safe, it can perform double computation of assignment and authorized verifiers
+	// It is safe in regards that only one collector will be stored to the cache
+	// In terms of locking time it's better to perform extra computation in edge cases than lock this logic with mutex
+	// since it's quite unlikely that same incorporated result will be processed by multiple goroutines simultaneously
+
+	// chunk assigment is based on the first block in the fork that incorporates the result
+	assignment, err := rsr.assigner.Assign(incorporatedResult.Result, incorporatedBlockID)
+	if err != nil {
+		return fmt.Errorf("could not determine chunk assignment: %w", err)
+	}
+
+	incorporatedBlock, err := rsr.headers.ByBlockID(incorporatedBlockID)
+	if err != nil {
+		return fmt.Errorf("could not determine height of incorporated block %s: %w",
+			incorporatedBlockID, err)
+	}
+
+	collector := NewApprovalCollector(incorporatedResult, incorporatedBlock, assignment, rsr.seals, rsr.requiredApprovalsForSealConstruction)
+
+	rsr.putCollector(incorporatedBlockID, collector)
```

Comment:
this method would override an approval collector, which we had already stored. 
How about
```golang
func (rsr *AssignmentCollector) putCollector(incorporatedBlockID flow.Identifier, collector *ApprovalCollector) bool {
	rsr.lock.Lock()
	defer rsr.lock.Unlock()
	if _, found := rsr.collectors[incorporatedBlockID]; found {
		return false
	}
	rsr.collectors[incorporatedBlockID] = collector
	return true
}
```
and then we can bail here, 
```suggestion
	isDuplicate := rsr.putCollector(incorporatedBlockID, collector)
	if isDuplicate {
		return nil
	}
```